### ISSUE TITLE
Links on their own line act as inclusiveLeft

### ISSUE THREAD

**Issue by wes-r** [ProseMirror/prosemirror#574]
Browser: Chrome v55
OS: Mac v10.11.6

Links on their own line act as if they are **inclusiveLeft** in that any typing at the beginning will be part of the link.

Steps to reproduce:
* Go to demo: http://prosemirror.net/demo/basic.html
* Create a link on its own line
* Place cursor at beginning of line
* type text
* BUG: You can see the text briefly flash without being a link and then "join" the link to the right

**Comment by marijnh** [comment]
Yes, this is how that behavior is intentionally defined. Is this bug a way of asking for a possibility to change that?

**Comment by wes-r** [comment]
Well I guess so :). The jury is still out if we definitely want to go with making links **not** inclusiveLeft. It's what we currently have and this breaks consistency/expectations (for us and for ProseMirror defaults). At the very least I'd like to know hot to override it.

**Comment by marijnh** [comment]
I'm tempted to rename `inclusiveLeft` to `inclusive`, and make the existing left-inclusive behavior, and the question of whether to be inclusive when at the start of a textblock, controllable by a single option. Does that sound like it would cover your intention for link behavior?

**Comment by wes-r** [comment]
Yes, although having both `inclusive` and `inclusiveRight` is confusing.

**Comment by marijnh** [comment]
Yes, that's why I said _rename_. I.e. `inclusive` would imply the current `inclusiveRight` behavior, and `inclusiveRight` would no longer exist. Does that still sound okay?

**Comment by wes-r** [comment]
Hmmm.. it sounds like there's three options:
* inclusiveRight
* inclusiveLeft
* inclusiveAtStart

Not sure what the use cases are for `inclusveAtStart` but I *may* want to have links be inclusiveLeft BUT NOT inclusiveRight.

Perhaps we should start a discussion in the forum to see other peoples use cases?

**Comment by marijnh** [comment]
You'd want links to behave exactly opposite from other marks (where you get the mark from the text before the cursor)? Are you sure? Three different options sounds like too much complexity for this.

**Comment by wes-r** [comment]
I don't think users think of links as the same as bold/italics/etc. I agree three seem like a lot. What is the reasoning behind the `inclusiveAtStart` functionality? Why the special case?

And my thinking for links being inclusive left is that if a user clicks to the left of a link, why not let it add to the link. BUT if the user finishes creating a link (and their cursor is to the right of the link) then you don't want them hitting space and typing to be part of the link. As of right now, I have links as non inclusive (left and right).

**Comment by marijnh** [comment]
> Why the special case?

When you insert text, the editor has to figure out where to take the marks from. If you're at a boundary between differently marked kinds of text, it prefers the text before (which you can override with `inclusiveRight`), if there's only text on one side (i.e. at the start of a textblock), it simply uses that text's marks.

> if a user clicks to the left of a link, why not let it add to the link.

Because the general expectation is that when you type, you get the style from the left of your cursor.

**Comment by wes-r** [comment]
> Because the general expectation is that when you type, you get the style from the left of your cursor.

Yes, but that is annoying when it comes to links! I think your proposal of having a single `inclusive` property will satisfy my current requirements. I also just checked and google docs treats links as non inclusive (so `inclusive: false` would yield google doc behavior for links).

**Comment by CMTietgen** [comment]
A similar issue happens when you create a link at the beginning of a line, then backspace to erase the link, then start typing some new text? When you do this the next text is changed into a link, but I would expect it to be plain text, since I just erased the link I added.

![prosemirror](https://cloud.githubusercontent.com/assets/13341574/23725206/aae9057a-041d-11e7-884b-8b8b37e58307.gif)

I compared how ProseMirror works to how Hotmail's rich editor works and the Hotmail editor switches to plain text if you erase the link to the beginning of the line.
